File: 05-2.rb
#!/usr/bin/env ruby

class Rule
  attr_reader :from, :to

  def initialize(line)
    @from, @to = line.split('|').map(&:to_i)
  end

  def inspect
    to_s
  end

  def to_s
    "<#{self.class}: #{@from} -> #{@to}>"
  end
end

class Update
  attr_reader :pages

  def initialize(line, rules)
    @pages = line.split(',').map(&:to_i)
    @rules = rules
  end

  def satisfy?(rule)
    first = @pages.find_index rule.from
    last = @pages.find_index rule.to

    return true if first.nil? or last.nil?

    first < last
  end

  def sort!
    @pages.sort! do |a, b|
      if @rules.any? { |r| r.from == a and r.to == b }
        -1
      elsif @rules.any? { |r| r.from == b and r.to == a }
        1
      else
        0
      end
    end
  end

  def score
    @pages[@pages.size / 2]
  end

  def inspect
    to_s
  end

  def to_s
    "<#{self.class}: #{@pages}>"
  end
end

class Puzzle
  def initialize(input)
    rules, updates = input.partition { |l| l.include? '|' }

    @rules = rules.map { |r| Rule.new r }
    @updates = updates.reject { |u| u == '' }.map { |u| Update.new(u, @rules) }
  end

  def score
    @updates.map do |u|
      if not @rules.all? { |r| u.satisfy? r }
        u.sort!
        u.score
      else
        0
      end
    end.sum
  end
end

input = File.read('05.input').lines.map(&:strip)
puzzle = Puzzle.new(input)
puts puzzle.score

File: day-05/src/05-2.rs
use std::cmp::Ordering;
use std::fs;

#[derive(Debug)]
struct Rule {
    from: i32,
    to: i32,
}

impl Rule {
    fn new(data: &str) -> Rule {
        let terms: Vec<i32> = data.split('|').map(|x| x.parse().unwrap()).collect();

        Rule {
            from: terms[0],
            to: terms[1],
        }
    }
}

impl Clone for Rule {
    fn clone(&self) -> Rule {
        Rule {
            from: self.from,
            to: self.to,
        }
    }
}

#[derive(Debug)]
struct Update {
    pages: Vec<i32>,
    rules: Vec<Rule>,
}

impl Update {
    fn new(data: &str, rules: &Vec<Rule>) -> Update {
        let pages: Vec<i32> = data.split(',').map(|x| x.parse().unwrap()).collect();

        Update {
            pages,
            rules: rules.to_vec(),
        }
    }

    fn satisfies(&self, rule: &Rule) -> bool {
        let first = match self.pages.iter().position(|p| p == &rule.from) {
            None => { return true }
            Some(x) => x
        };

        let last = match self.pages.iter().position(|p| p == &rule.to) {
            None => { return true }
            Some(x) => x
        };

        first < last
    }

    fn sort(&mut self) {
        self.pages.sort_by(|a, b| {
            if self.rules.iter().any(|r| r.from == *a && r.to == *b) {
                Ordering::Less
            } else if self.rules.iter().any(|r| r.from == *b && r.to == *b) {
                Ordering::Greater
            } else {
                Ordering::Equal
            }
        });
    }

    fn score(&self) -> i32 {
        self.pages[self.pages.len() / 2]
    }
}

#[derive(Debug)]
struct Puzzle {
    rules: Vec<Rule>,
    updates: Vec<Update>,
}

impl Puzzle {
    fn new(data: &str) -> Puzzle {
        let (rules, updates): (Vec<_>, Vec<_>) =
            data
            .lines()
            .filter(|l| !l.is_empty())
            .partition(|l| l.contains('|'));

        let rules = rules.iter().map(|r| Rule::new(r)).collect();
        let updates = updates.iter().map(|u| Update::new(u, &rules)).collect();

        Puzzle {
            rules,
            updates,
        }
    }

    fn score(&mut self) -> i32 {
        self.updates
            .iter_mut()
            .map(|u| {
                match self.rules.iter().all(|r| u.satisfies(r)) {
                    true => 0,
                    false => {
                        u.sort();
                        u.score()
                    }
                }
            })
            .sum()
    }
}

fn main() {
    let input = fs::read_to_string("../05.input").unwrap();

    let mut p = Puzzle::new(&input);

    println!("{}", p.score());
}
